File: stockify/releases/v0.2.0/packages/core/src/providers/github.ts
import type { Candidate, FeedSource } from "../types.ts";
import { stableId } from "../utils.ts";

type GitHubSource = Extract<FeedSource, {type:"github_search"}>;

export async function fetchGitHub(source: GitHubSource, env: { githubToken?: string; githubBaseUrl?: string }): Promise<Candidate[]> {
  const base=(env.githubBaseUrl||"https://api.github.com").replace(/\/$/,"");
  const params=new URLSearchParams({q:source.query,sort:source.sort||"updated",order:source.order||"desc",per_page:"30"});
  const headers:Record<string,string>={"accept":"application/vnd.github+json","user-agent":"Feedify/0.2 (+https://feedify.dev)","x-github-api-version":"2022-11-28"};
  if(env.githubToken)headers.authorization=`Bearer ${env.githubToken}`;
  const res=await fetch(`${base}/search/repositories?${params}`,{headers,signal:AbortSignal.timeout(12000)});
  if(!res.ok)throw new Error(`GitHub search ${res.status}: ${await res.text()}`);
  const json:any=await res.json();
  return (Array.isArray(json.items)?json.items:[]).map((repo:any)=>({
    id:stableId(source.id,String(repo.id||repo.full_name)),sourceId:source.id,sourceType:"github_search" as const,sourceLabel:source.label||"GitHub",
    title:repo.full_name,text:[repo.description,`★ ${repo.stargazers_count||0}`,repo.language?`Language: ${repo.language}`:null].filter(Boolean).join(" · "),
    url:repo.html_url,author:repo.owner?.login,publishedAt:repo.updated_at,engagement:Number(repo.stargazers_count||0),metadata:{stars:repo.stargazers_count,forks:repo.forks_count,language:repo.language,topics:repo.topics}
  })).filter((x:any)=>x.text);
}

File: stockify/releases/v0.3.0-alpha/feedify/adapters/github.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone

from feedify.schemas import NormalizedItem
from feedify.settings import get_settings

from .base import SourceAdapter
from .utils import parse_datetime


class GitHubAdapter(SourceAdapter):
    name = "github"
    base_url = "https://api.github.com"

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.token = get_settings().github_token

    async def fetch(self, limit: int = 25) -> list[NormalizedItem]:
        since = (datetime.now(timezone.utc) - timedelta(days=21)).date().isoformat()
        q = f"created:>{since} stars:>15"
        headers = {"Accept": "application/vnd.github+json", "User-Agent": "feedify-alpha/0.1"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        response = await self.client.get(
            f"{self.base_url}/search/repositories",
            headers=headers,
            params={"q": q, "sort": "stars", "order": "desc", "per_page": min(limit, 100)},
        )
        response.raise_for_status()
        rows = response.json().get("items", [])
        return [
            NormalizedItem(
                source_type=self.name,
                external_id=row["full_name"],
                title=row["full_name"],
                url=row.get("html_url"),
                body=row.get("description"),
                author=(row.get("owner") or {}).get("login"),
                published_at=parse_datetime(row.get("created_at")),
                metrics={
                    "stars": row.get("stargazers_count") or 0,
                    "forks": row.get("forks_count") or 0,
                    "watchers": row.get("watchers_count") or 0,
                    "open_issues": row.get("open_issues_count") or 0,
                    "language": row.get("language"),
                    "topics": row.get("topics") or [],
                    "pushed_at": row.get("pushed_at"),
                },
                raw=row,
            )
            for row in rows
        ]
